packages/@react-aria/focus/src/FocusScope.tsx

```diff
@@ -458,12 +477,30 @@ function useRestoreFocus(scopeRef: RefObject<HTMLElement[]>, restoreFocus: boole
         document.removeEventListener('keydown', onKeyDown, true);
       }
 
-      if (restoreFocus && nodeToRestore && isElementInScope(document.activeElement, scopeRef.current)) {
-        requestAnimationFrame(() => {
-          if (document.body.contains(nodeToRestore)) {
-            focusElement(nodeToRestore);
-          }
-        });
+      if (nodeToRestore) {
+        let index = nodeToRestoreArray.indexOf(nodeToRestore);
+
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+        if (restoreFocus && isElementInScope(document.activeElement, scopeRef.current)) {
+          requestAnimationFrame(() => {
+            if (document.body.contains(nodeToRestore) && nodeToRestore !== document.body) {
+              focusElement(nodeToRestore);
+            } else {
+              while (index > 0) {
+                index--;
+                let node = nodeToRestoreArray[index];
```

Comment:
can there be items at indices greater than index? i think it'd be better if we actually treated nodeToRestoreArray as a stack and used `Array.prototype.pop()`
I wouldn't provide a method to remove certain indices either. I'd just pop, if it's not in the dom, continue popping.

otherwise it looks to me like there are some times where the nodeToRestore maybe isn't removed from the array? such as in the if statement on line 486?

Reply:
Yes there can. I think this is what happens when a dialog opens from a menu item and the menu then closes. The menu item gets added to the end of the nodeToRestoreArray, to serve as the node to restore for the dialog. Then the FocusScope for the menu unmounts removing the nodeToRestore for the Dialog from the DOM. We need to clean up the array, removing the node that is no longer present in the DOM. 

See this example: https://reactspectrum.blob.core.windows.net/reactspectrum/624fe15622aaea936834c922cb308877c83b3d83/storybook/index.html?path=/story/dialogcontainer--nested-dialog-containers

Reply:
but when we're popping off items in the stack, won't we check if they are in the dom? so we don't need to clear them out ahead of time?

Reply:
I think I may now have a fix for this: https://github.com/adobe/react-spectrum/pull/2445/commits/a2fad80eaffb0724db6b247daefb0e118aea9a51